src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala

```diff
@@ -254,7 +258,9 @@ class MarathonSchedulerService @Inject() (
           electionService.abdicateLeadership(error = result.isFailure, reoffer = isRunningLatch.getCount > 0)
         }
 
-        driver = None
+        synchronized {
+          driver = None
+        }
 
         log.info(s"Call postDriverRuns callbacks on ${prePostDriverCallbacks.mkString(", ")}")
         Await.result(Future.sequence(prePostDriverCallbacks.map(_.postDriverTerminates)), config.zkTimeoutDuration)
```

Comment:
I'm sensing there might be a race condition with driver being accessed outside of synchronized. Could we have the synchronized block return a boolean whether or not driver was nonEmpty going into the synchronized block, and then call electionService.abdicateLeadership after?

Something like this?

```diff
diff --git a/src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala b/src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala
index ea32ca7..d8075ec 100644
--- a/src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala
+++ b/src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala
@@ -237,7 +237,7 @@ class MarathonSchedulerService @Inject() (
         driver.foreach(_.run())
       }
     } onComplete { result =>
-      synchronized {
+      {
 
         log.info(s"Driver future completed with result=$result.")
         result match {
@@ -245,17 +245,21 @@ class MarathonSchedulerService @Inject() (
           case _ =>
         }
 
+        val driverCleared = synchronized {
+          val cleared = driver.nonEmpty
+          driver = None
+          cleared
+        }
+
         // ONLY do this if there's some sort of driver crash: avoid invoking abdication logic if
         // the driver was stopped via stopDriver. stopDriver only happens when
         //   1. we're being terminated (and have already abdicated)
         //   2. we've lost leadership (no need to abdicate if we've already lost)
-        driver.foreach { _ =>
+        if (driverCleared) {
           // tell leader election that we step back, but want to be re-elected if isRunning is true.
           electionService.abdicateLeadership(error = result.isFailure, reoffer = isRunningLatch.getCount > 0)
         }
 
-        driver = None
-
         log.info(s"Call postDriverRuns callbacks on ${prePostDriverCallbacks.mkString(", ")}")
         Await.result(Future.sequence(prePostDriverCallbacks.map(_.postDriverTerminates)), config.zkTimeoutDuration)
         log.info(s"Finished postDriverRuns callbacks")
```